### ISSUE TITLE
EagerTensor API extensions for tensor4all-rs migration

### ISSUE THREAD

**Issue by shinaoka** [tensor4all/tenferro-rs#706]
## Summary

tensor4all-rs is migrating to use tenferro's `EagerTensor` as its core tensor type. The current `EagerTensor` only supports 6 operations (add, mul, neg, exp, reduce_sum, dot_general). This issue tracks the extensions needed before tensor4all-rs can adopt `EagerTensor`.

Design doc: `tensor4all-rs/docs/design/symbolic-shape-and-tenferro-migration.md` (2026-04-13 revision)

## Motivation

tensor4all-rs will use two tenferro types:
- **`EagerTensor`** for `TensorDynLen` (indexed tensors, dynamic dtype, optional AD)
- **`TypedTensor<T>`** for SimpleTT `Tensor<T,N>` (no indices, static dtype, no AD)

When `requires_grad=false` (default), `EagerTensor` creates zero `GradNode`s — primal-only computation with no overhead. AD is opt-in.

#### 1. Multi-output op pattern for EagerTensor

Current `EagerTensor` only supports single-output ops via `unary_op` / `binary_op`. Linalg decompositions return multiple tensors (SVD → 3, QR → 2, LU → 4). Need:

- `exec_multi_output()` integration (already exists in `eager_exec.rs`)
- Each output gets its own `EagerTensor` with shared `GradNode` but different `output_idx`
- When `requires_grad=false`, no `GradNode` created (zero overhead)

Methods to add:

| Method | Returns | StdTensorOp |
|--------|---------|-------------|
| `svd()` | `(Self, Self, Self)` | `SVD` |
| `qr()` | `(Self, Self)` | `QR` |
| `lu()` | `(Self, Self, Self, Self)` | `LU` |
| `cholesky()` | `Self` | `Cholesky` |
| `eigh()` | `(Self, Self)` | `Eigh` |
| `eig()` | `(Self, Self)` | `Eig` |
| `triangular_solve(b, ...)` | `Self` | `TriangularSolve` |

#### 2. Einsum on EagerTensor

N-ary contraction with AD support:

```rust
pub fn eager_einsum_ad<B: TensorBackend>(
    inputs: &[&EagerTensor<B>],
    subscripts: &str,
) -> EagerTensor<B>
```

- Internally calls existing `eager_einsum(ctx, tensors, subscripts)`
- Builds `GradNode` if any input has `requires_grad=true`
- Merges `EagerContext`s from all inputs
- AD rule: VJP w.r.t. input i = einsum of cotangent with remaining inputs

#### 3. Structural ops on EagerTensor

Wrap existing `StdTensorOp` variants as methods:

| Method | Signature | StdTensorOp |
|--------|-----------|-------------|
| `transpose` | `(&self, perm: &[usize]) -> Self` | `Transpose` |
| `reshape` | `(&self, shape: &[usize]) -> Self` | `Reshape` |
| `slice` | `(&self, starts, limits, strides) -> Self` | `Slice` |
| `concatenate` | `(tensors: &[&Self], axis) -> Self` | `Concatenate` |
| `broadcast_in_dim` | `(&self, shape, dims) -> Self` | `BroadcastInDim` |
| `pad` | `(&self, value, low, high, interior) -> Self` | `Pad` |
| `reverse` | `(&self, axes) -> Self` | `Reverse` |
| `gather` | `(&self, indices, ...) -> Self` | `Gather` |
| `dynamic_slice` | `(&self, starts, sizes) -> Self` | `DynamicSlice` |

#### 4. EagerContext public constructor + shared context support

Current `EagerContext` is private. tensor4all-rs needs a thread-local shared context to avoid repeated `absorb_from()` on every binary op:

```rust
impl<B: TensorBackend> EagerContext<B> {
    pub fn with_backend(backend: B) -> Rc<Self>;
}

impl<B: TensorBackend> EagerTensor<B> {
    pub fn from_tensor_in(tensor: Tensor, ctx: Rc<EagerContext<B>>) -> Self;
    pub fn requires_grad_in(tensor: Tensor, ctx: Rc<EagerContext<B>>) -> Self;
}
```

#### 5. Elementwise ops on EagerTensor

| Method | StdTensorOp |
|--------|-------------|
| `div`, `abs`, `conj`, `sign` | `Div`, `Abs`, `Conj`, `Sign` |
| `log`, `sqrt`, `sin`, `cos`, `tanh`, `pow` | `Log`, `Sqrt`, `Sin`, `Cos`, `Tanh`, `Pow` |
| `select`, `maximum`, `minimum` | `Select`, `Maximum`, `Minimum` |
| `expm1`, `log1p`, `rsqrt` | `Expm1`, `Log1p`, `Rsqrt` |

#### 6. Diagonal ops on EagerTensor

| Method | StdTensorOp |
|--------|-------------|
| `extract_diag` | `ExtractDiag` |
| `embed_diag` | `EmbedDiag` |
| `tril`, `triu` | `Tril`, `Triu` |

#### 7. Additional reduction ops

| Method | StdTensorOp |
|--------|-------------|
| `reduce_prod` | `ReduceProd` |
| `reduce_max` | `ReduceMax` |
| `reduce_min` | `ReduceMin` |

#### 8. TypedTensor convenience methods (optional)

Convenience methods on `TypedTensor<T>` wrapping free functions:

```rust
impl<T: TensorScalar> TypedTensor<T> {
    pub fn einsum(ctx: &mut CpuBackend, inputs: &[&Self], subscripts: &str) -> Self;
    pub fn svd(&self, ctx: &mut CpuBackend) -> (Self, TypedTensor<T::Real>, Self);
    pub fn qr(&self, ctx: &mut CpuBackend) -> (Self, Self);
}
```

## Implementation Notes

All operations in P0/P1 already exist as `StdTensorOp` variants with `TensorBackend` implementations. The work is:

1. Add method on `EagerTensor` that constructs the `StdTensorOp`
2. Call `exec_single_output()` or `exec_multi_output()`
3. Build `GradNode` if `requires_grad=true` (following existing `unary_op`/`binary_op` pattern)
4. AD rules (linearize/transpose) already exist for most ops in `tenferro-ops`

The multi-output pattern (item 1) is the only structural addition; everything else follows existing patterns.

**Comment by shinaoka** [comment]
P2 (TypedTensor convenience methods) is also in scope for this issue.

**Comment by shinaoka** [comment]
Phase 1–5 implemented by #709 and #710 (both merged), including P2 (TypedTensor convenience methods). Keeping open until all checklist items are verified against the current codebase.
